#
# How to apply this patch to vanilla Linux linux-2.6.28 kernel source tree:
#
#	# cd /usr/src/your/linux/kernel/path/.../drivers/input/mouse
#	# patch -p1 < /path/to/this/patch/sentelic-2.6.28.patch
#	# cp /path/to/this/patch/sentelic.[ch] .
#	(then follow your regular kernel building procedure...)
#
--- mouse.orig/Makefile	2009-01-11 21:42:21.000000000 +0800
+++ mouse/Makefile	2009-01-11 21:46:04.000000000 +0800
@@ -18,7 +18,7 @@ obj-$(CONFIG_MOUSE_HIL)		+= hil_ptr.o
 obj-$(CONFIG_MOUSE_VSXXXAA)	+= vsxxxaa.o
 obj-$(CONFIG_MOUSE_GPIO)	+= gpio_mouse.o
 
-psmouse-objs := psmouse-base.o synaptics.o
+psmouse-objs := psmouse-base.o synaptics.o sentelic.o
 
 psmouse-$(CONFIG_MOUSE_PS2_ALPS)	+= alps.o
 psmouse-$(CONFIG_MOUSE_PS2_ELANTECH)	+= elantech.o
--- mouse.orig/psmouse.h	2009-01-11 21:42:27.000000000 +0800
+++ mouse/psmouse.h	2009-01-11 21:45:42.000000000 +0800
@@ -91,6 +91,7 @@ enum psmouse_type {
 	PSMOUSE_CORTRON,
 	PSMOUSE_HGPK,
 	PSMOUSE_ELANTECH,
+	PSMOUSE_FSP,
 	PSMOUSE_AUTO		/* This one should always be last */
 };
 
--- mouse.orig/psmouse-base.c	2009-01-11 21:42:34.000000000 +0800
+++ mouse/psmouse-base.c	2009-01-11 21:45:31.000000000 +0800
@@ -30,6 +30,7 @@
 #include "trackpoint.h"
 #include "touchkit_ps2.h"
 #include "elantech.h"
+#include "sentelic.h"
 
 #define DRIVER_DESC	"PS/2 mouse driver"
 
@@ -639,6 +640,20 @@ static int psmouse_extensions(struct psm
 	}
 
 /*
+ * Try Finger Sensing Pad
+ */
+	if (max_proto > PSMOUSE_IMEX) {
+		if (fsp_detect(psmouse, set_properties) == 0) {
+			if (!set_properties || fsp_init(psmouse) == 0)
+				return PSMOUSE_FSP;
+/*
+ * Init failed, try basic relative protocols
+ */
+ 			max_proto = PSMOUSE_IMEX;
+		}
+	}
+
+/*
  * Try OLPC HGPK touchpad.
  */
 	if (max_proto > PSMOUSE_IMEX &&
@@ -813,6 +828,13 @@ static const struct psmouse_protocol psm
 	},
  #endif
 	{
+		.type		= PSMOUSE_FSP,
+		.name		= "FSPPS/2",
+		.alias		= "fsp",
+		.detect		= fsp_detect,
+		.init		= fsp_init,
+	},
+	{
 		.type		= PSMOUSE_CORTRON,
 		.name		= "CortronPS/2",
 		.alias		= "cortps",
